# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/PlainObject.h
# Commit: bad57461dcc4
# Full Hash: bad57461dcc4c6c52064bf985ee737d8abbc183b
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-23 21:43:59
# Regressor Bug: 1696178
# File Overlap Count: 1
# Description:
#   Bug 1696178 part 5 - Add ObjectFlag::HasNonWritableOrAccessorExcludingProto. r=jonco
#   
#   This will be used by the next patch to optimize Object.assign.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D108407
# ==============================================================================

diff -r 12574c8b209d -r bad57461dcc4 js/src/vm/PlainObject.h
--- a/js/src/vm/PlainObject.h	Tue Mar 23 12:43:56 2021 +0000
+++ b/js/src/vm/PlainObject.h	Tue Mar 23 12:43:57 2021 +0000
@@ -26,11 +26,27 @@
  public:
   static const JSClass class_;
 
+ private:
+#ifdef DEBUG
+  void assertHasNoNonWritableOrAccessorPropExclProto() const;
+#endif
+
+ public:
   static inline JS::Result<PlainObject*, JS::OOM> createWithTemplate(
       JSContext* cx, JS::Handle<PlainObject*> templateObject);
 
   /* Return the allocKind we would use if we were to tenure this object. */
   inline gc::AllocKind allocKindForTenure() const;
+
+  bool hasNonWritableOrAccessorPropExclProto() const {
+    if (hasFlag(ObjectFlag::HasNonWritableOrAccessorPropExclProto)) {
+      return true;
+    }
+#ifdef DEBUG
+    assertHasNoNonWritableOrAccessorPropExclProto();
+#endif
+    return false;
+  }
 };
 
 // Specializations of 7.3.23 CopyDataProperties(...) for NativeObjects.